### ISSUE TITLE
Higher precision decimals

### ISSUE THREAD

**Issue by maciejhirsz** [maciejhirsz/json-rust#79]
So I'm going back and forth on the idea of [using a custom number type](https://github.com/maciejhirsz/json-rust/issues/60).

While it's nice to be able to parse higher precision than `f64`, it's also pretty useless unless there is a way to use that somehow. A solution for that could be methods like:
- `as_fixed_point_u64(u16) -> Option<u64>`
- `as_fixed_point_i64(u16) -> Option<i64>`

This way a number such as `12.3456` can be obtained as an integer `123456` when called with fixed point argument `4`. A way to construct such numbers would be needed as well (macro?)

Pros:
- Much faster serializing (canada.json will be a breeze).
- Parsing higher precision might be actually useful.

Cons:
- Obscuring the API, though this is already done with `Object` and `Short` types.
- Converting floats to the decimal representation is costly. I've already done an experimental alteration of the dtoa crate to produce decimal mantissa and exponent instead of writing to string with the grisu2 algorithm. Pefromance of that is quite okay.

Open questions:
- Is 64 bit mantissa for precision enough?
- Do we enable maths on the `Number` type and/or `JsonValue`?
- Can we use feature flags to convert numer types from other crates, in particular `num` crate?

**Comment by Mark-Simulacrum** [comment]
> Converting floats to the decimal representation is costly. I've already done an experimental alteration of the dtoa crate to produce decimal mantissa and exponent instead of writing to string with the grisu2 algorithm. Performance of that is quite okay.

I'm not sure I follow you: Is the conversion costly, but acceptable, or is the above discussing the conversion into and out of the custom number type? Converting into it is costly, converting out is decently fast?

**Comment by maciejhirsz** [comment]
Basically, from my benchmarks on DEC64 using `3.141592653589793` on my hardware:

- Printing out of a float is something along the lines of 120ns (that's using the grisu2 implemented in dtoa). This is pretty much the [fastest way to do it](http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf).
- If stored as an integer with decimal exponent, printing out is slightly under 40ns. I've been tweaking this a lot.
- Converting an f64 to an integer mantissa + decimal exponent is roughly 80ns using a variation on grisu2.
- Combined cost of converting to decimal and printing from decimal is comparable to just printing out the float. In some cases I even got better results with this intermediate conversion, but not often.

(Note: when I say "float" I mean binary float, technically they are both floats, one is binary the other is decimal, but I can't unlearn saying "float" to mean "binary floating point")

The main pain point here is that if you do want to do arithmetic on `f64`, you will be taxed every time you want to store your result as `JsonValue`. If you do it just once, the cost is negligible (you get it back when serializing), otherwise you do extra work on each new assignment that you wouldn't have if the internal representation was plain `f64`.

And yes, it's only `f64` -> decimal that is costly. The other way around is trivial (2 nanos or so), since you can use cached powers of 10 to convert whatever decimal exponent to fraction. I'm already [doing that when parsing numbers](https://github.com/maciejhirsz/json-rust/blob/bf0e27ca3abf66a2aedbab91512f2a61026e8283/src/parser.rs#L249).

**Comment by maciejhirsz** [comment]
So, I've (once again) ripped away stuff from what I did for DEC64 before.

0.9.1 on my hardware:

```
======= json-rust ======== parse|stringify === parse|stringify ===
data/canada.json          17.5ms    16.8ms
data/citm_catalog.json    10.1ms     1.4ms
data/twitter.json          3.6ms     1.0ms
```

0.10.0 on my hardware:

```
======= json-rust ======== parse|stringify === parse|stringify ===
data/canada.json          17.3ms     5.3ms
data/citm_catalog.json    10.5ms     1.3ms
data/twitter.json          3.6ms     0.9ms
```

**Comment by dtolnay** [comment]
There are some concerning test failures though:

```
assertion failed: `(left == right)` (left: `"3.141592653589793"`, right: `"3.141592653589793e-16"`)
assertion failed: `(left == right)` (left: `"1"`, right: `"10000000000000000000"`)
assertion failed: `(left == right)` (left: `Number(Number { positive: true, mantissa: 5, exponent: 2 })`, right: `0.05`)
```

It may not be printing what you think it is.

**Comment by maciejhirsz** [comment]
Yeah, I'm quite aware, my own tests are failing on this. At this moment it's only managing integers and fractions without  within an arbirtrary "printable range". That is, a number with any positive exponent or negative exponent below -17 will print incorrectly.

I only pushed it to see the benchmark because (as far as I've looked over the file) `canada.json` floats are within what I can print already.

**Comment by maciejhirsz** [comment]
@dtolnay how did you produce this one:

```
assertion failed: `(left == right)` (left: `Number(Number { positive: true, mantissa: 5, exponent: 2 })`, right: `0.05`)
```

Exponent should be -2, both parsing from string and converting from float I get it correctly.

**Comment by dtolnay** [comment]
Just running `cargo test` on the current 0.10.0 branch (8dbbb93). Here is the full output:

```
failures:

---- unit::parse_number_with_e stdout ----
	thread 'unit::parse_number_with_e' panicked at 'assertion failed: `(left == right)` (left: `Number(Number { positive: true, mantissa: 5, exponent: 2 })`, right: `500`)', tests/lib.rs:392
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- unit::parse_number_with_negative_e stdout ----
	thread 'unit::parse_number_with_negative_e' panicked at 'assertion failed: `(left == right)` (left: `Number(Number { positive: true, mantissa: 5, exponent: 2 })`, right: `0.05`)', tests/lib.rs:404

---- unit::parse_very_long_float stdout ----
	thread 'unit::parse_very_long_float' panicked at 'assertion failed: `(left == right)` (left: `Number(Number { positive: true, mantissa: 2225073858507201136, exponent: 18308 })`, right: `0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002225073858507201`)', tests/lib.rs:362

---- unit::parse_number_with_positive_e stdout ----
	thread 'unit::parse_number_with_positive_e' panicked at 'assertion failed: `(left == right)` (left: `Number(Number { positive: true, mantissa: 5, exponent: 2 })`, right: `500`)', tests/lib.rs:398

---- unit::stringify_large_number stdout ----
	thread 'unit::stringify_large_number' panicked at 'assertion failed: `(left == right)` (left: `"1"`, right: `"10000000000000000000"`)', tests/lib.rs:181

---- unit::stringify_very_small_number stdout ----
	thread 'unit::stringify_very_small_number' panicked at 'assertion failed: `(left == right)` (left: `"3.141592653589793"`, right: `"3.141592653589793e-16"`)', tests/lib.rs:191

---- unit::stringify_very_large_number stdout ----
	thread 'unit::stringify_very_large_number' panicked at 'assertion failed: `(left == right)` (left: `"3141592653589793"`, right: `"3.141592653589793e50"`)', tests/lib.rs:186


failures:
    unit::parse_number_with_e
    unit::parse_number_with_negative_e
    unit::parse_number_with_positive_e
    unit::parse_very_long_float
    unit::stringify_large_number
    unit::stringify_very_large_number
    unit::stringify_very_small_number

test result: FAILED. 129 passed; 7 failed; 0 ignored; 0 measured
```

**Comment by maciejhirsz** [comment]
6 to go...

**Comment by maciejhirsz** [comment]
Tests fixed, some more added, still 5.3ms :)
